vm/src/warn.rs

```diff
@@ -31,19 +32,119 @@ pub fn warn(
+    //         return NULL;
+    // }
+    // else {
+    //     Py_SETREF(st->filters, warnings_filters);
+    // }
+
+    let filters = vm.ctx.types.filter_type.as_object().to_owned();
+    let filters = PyListRef::try_from_object(vm, filters);
+    if filters.is_err() {
+        return Err(vm.new_value_error("_warnings.filters must be a list".to_string()));
+    }
+
+    Ok(())
+}
+
+fn already_warned(
+    registry: PyObjectRef,
+    key: PyObjectRef,
+    _should_set: usize,
+    vm: &VirtualMachine,
+) -> bool {
+    let _version_obj = registry.get_item("version", vm);
+    let already_warned = registry.get_item(key.as_ref(), vm);
+
+    // if (version_obj == NULL
+    //     || !PyLong_CheckExact(version_obj)
+    //     || PyLong_AsLong(version_obj) != st->filters_version)
+    // {
+    //     if (PyErr_Occurred()) {
+    //         return -1;
+    //     }
+    //     PyDict_Clear(registry);
+    //     version_obj = PyLong_FromLong(st->filters_version);
+    //     if (version_obj == NULL)
+    //         return -1;
+    //     if (PyDict_SetItem(registry, &_Py_ID(version), version_obj) < 0) {
+    //         Py_DECREF(version_obj);
+    //         return -1;
+    //     }
+    //     Py_DECREF(version_obj);
+    // }
+    // else {
+    already_warned.is_ok()
+    // }
+
+    /* This warning wasn't found in the registry, set it. */
+    // if should_set != 0 {
+    //     registry.set_item(key.as_ref(), vm.ctx.new_bool(true).into(), vm);
+    // }
+
+    // return 0
+}
+
+fn normalize_module(filename: PyStrRef, vm: &VirtualMachine) -> Option<PyObjectRef> {
+    let len = filename.char_len();
+
+    if len == 0 {
+        Some(vm.new_pyobj("<unknown>"))
+    } else if len >= 3 && filename.as_str().contains(".py") {
+        Some(vm.new_pyobj(filename.as_str().replace(".py", "")))
+    } else {
+        Some(filename.as_object().to_owned())
+    }
+}
+
 #[allow(clippy::too_many_arguments)]
 fn warn_explicit(
     category: Option<PyTypeRef>,
     message: PyStrRef,
-    _filename: PyStrRef,
-    _lineno: usize,
-    _module: PyObjectRef,
-    _registry: PyObjectRef,
+    filename: PyStrRef,
+    lineno: usize,
+    module: Option<PyObjectRef>,
+    registry: PyObjectRef,
     _source_line: Option<PyObjectRef>,
     _source: Option<PyObjectRef>,
     vm: &VirtualMachine,
 ) -> PyResult<()> {
-    // TODO: Implement correctly
+    // if registry.is(vm.ctx.types.none_type.as_object()) || !registry.dict() {
+    //     return Err(vm.new_type_error("'registry' must be a dict or None".to_string()));
+    // }
+
+    let module = module.and(normalize_module(filename, vm));
+
+    /* module can be None if a warning is emitted late during Python shutdown.
+    In this case, the Python warnings module was probably unloaded, filters
+    are no more available to choose as action. It is safer to ignore the
+    warning and do nothing. */
+    if module.is_none() {
+        // return Ok(&vm.ctx.types.none_type);
+        return Ok(());
+    }
```

Comment:
```suggestion
    let module = if let Some(module) = module {
        module
    } else {
        // return Ok(&vm.ctx.types.none_type);
        return Ok(());
    }
```

To let typesystem recognize module never be None later.